lib/collection/src/shards/local_shard.rs

```diff
@@ -492,12 +497,15 @@ impl LocalShard {
         // snapshot all shard's segment
         let snapshot_segments_shard_path = snapshot_shard_path.join("segments");
         create_dir_all(&snapshot_segments_shard_path).await?;
-        self.segments
-            .read()
-            .snapshot_all_segments(&snapshot_segments_shard_path)?;
 
-        // snapshot all shard's WAL
-        self.snapshot_wal(snapshot_shard_path)?;
+        {
+            // Do not change segments while snapshotting
+            let segments_read = self.segments.read();
+            segments_read.snapshot_all_segments(&snapshot_segments_shard_path)?;
+
+            // snapshot all shard's WAL
+            self.snapshot_wal(snapshot_shard_path)?;
+        }
```

Comment:
Hold `segments` read lock to ensure, that segments list is not modified during the snapshot (optimization does not interfere)